### ISSUE TITLE
Ad-hoc Collections in functions

### ISSUE THREAD

**Issue by larkob** [ArnoldSmith86/virtualtabletop#72]
Currently, functions allow you to work with collections or with content of holders. When you want to simply move or flip an individual card it seems a lot of overhead to first perform a SELECT. I propose to implement "ad-hoc collections" which would use jsonPath expressions as a means to allow for high flexibility while adhering to an established standard.

Example for moving a widget with specific ID to a holder:
```
{
	"func": “Move”,
        "collection": "$.[?@.id=='someID']"
	"to": ”<holderID>”
}
```

It has further advantages over Select operations in that you could for instance use multiple properties as search criteria. The operation to recall cards of a particular deck to a Holder could then look as follows:
```
{
	"func": “Move”
	"collection": "$.[?(@.type=='card' && @.deck=='sgmdD')]"
	"to": ”<holderID>”
}
```

Another potential advantage with jsonPath is that it would allow for picking the first / last / nth widget within a pile or holder, which would be more difficult to realize with a SELECT operation. Here’s an example of how to move just the first card:
```
{
	"func": “Move”
	"collection": "$.[?(@.type=='card' && @.deck=='sgmdD')][0]"
	"to": ”<holderID>”
}
```

Obviously, writing a jsonPath expression is not so easy, but once we’ve got Blockly implemented, expressions could easily be constructed with blocks.

**Comment by UltimateGeek** [comment]
Doing this with embedded javascript would require some sort of mechanism to sandbox the adhoc code in order to prevent code-injection attacks.

Implementing a sandbox may be possible (although technically challenging), but from a security standpoint, it would be best if we kept the capability of coding within vtt button actions to not involve evaluating custom javascript code.

I'd propose we do something like this:
* Add another property to Button to specify reusable routines, each of which can made up of any of the clickRoutine functions.
* Add DSL for collection references to invoke a reusable routine before using the collection.

```json
{
  "type": "button",
  "subroutines": {
    "selectThing": [
      { "input": "id" },
      { "func": "SELECT", "property": "type",
        "applyVariables": [ { "parameter": "value", "variable": "id" } ] 
      }
    ]
  },
  "clickRoutine": [
    { "func": "MOVE", "to": "aHolder",
      "collection": "DEFAULT@selectThing(an-id)" }
  ]
}
```

**Comment by larkob** [comment]
I agree, but it's not Javascript, and is therefore completely safe to use. I don't know if you're familiar with xPath or jquery selectors - it's actually modelled after xPath for XML and jsonPath implementations exist for several languages like JS, Python, Java, etc. The advantage of jsonPath expressions is that we don't have to invent and implement a DSL and can make use of existing JS libraries for jsonPath expressions.

**Comment by UltimateGeek** [comment]
> I agree, but it's not Javascript, and is therefore completely safe to use. I don't know if you're familiar with xPath or jquery selectors - it's actually modelled after xPath for XML and jsonPath implementations exist for several languages like JS, Python, Java, etc. The advantage of jsonPath expressions is that we don't have to invent and implement a DSL and can make use of existing JS libraries for jsonPath expressions.

Oh, that's jsonPath? Cool! Yeah, I'm well-versed in XPath, so I think using a jsonPath library is a great idea!

**Comment by rogerl50** [comment]
I'd love to see something like this. Don't know how many times I've looked at the `GET` code documentation only to realize that I need to create a collection first. (BTW, the one-liner at the start of the button documentation says this about GET: "GET - to get the property value of a widget". [not "a widgets in a collection."])

**Comment by larkob** [comment]
> BTW, the one-liner at the start of the button documentation says this about GET: "GET - to get the property value of a widget". [not "a widgets in a collection."])

Thanks, I've corrected it.

**Comment by 96LawDawg** [comment]
Closed by #881

**Comment by robartsd** [comment]
#881 provided basic functionality, but this PR requested more; still I think that if additional functionality is still needed we should open a new issue for it (perhaps referencing this one).
